This changes the PriorityScheduler high vs low priority task logic

jentfoo opened the pull request:
This changes the logic so that low priority tasks wont be executed if there are any high priority tasks UNLESS they have waited up to the max low priority wait time (assuming the high priority tasks are not being delayed that much, high priority always has priority).
It also changes from having two threads to consume tasks to just one.  This one thread now gets a worker first, then decides which task should be provided to that worker.  Because we get the worker first this resolves #75.
This does increase the jar size, and does not really change the performance (if anything it might be _slightly_ slower, but the benchmarks show a statistically insignificant difference).

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -600,15 +583,13 @@ protected void finalize() {
+    protected int currentPoolSize;  // is locked around workersLock
     private final AtomicBoolean shutdownStarted;
     private volatile boolean shutdownFinishing; // once true, never goes to false
-    private volatile long maxWaitForLowPriorityInMs;
```
Max wait is now handled in the queue handler, since it is determining what task should be run.  The worker pool only stores workers, and provides them as they are available.  We get the worker before we find out which task to run.

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -745,7 +722,12 @@ public int getCurrentPoolSize() {
     public int getCurrentRunningCount() {
       synchronized (workersLock) {
-        return currentPoolSize - availableWorkers.size();
+        int workerOutQty = currentPoolSize - availableWorkers.size();
```
This math changes because if there is a worker out (ie no thread blocked waiting for a worker), then it will be waiting in the consumer, not running a task.

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -883,138 +831,104 @@ protected void workerDone(Worker worker) {
-     * @throws InterruptedException Thrown if thread is interrupted while waiting for a worker
+     * @param task Task to find in queue and reposition based off next delay
      */
-    protected void runLowPriorityTask(TaskWrapper task) throws InterruptedException {
```
This is where the diff gets confusing...it's still showing removals of code that used to be in the "WorkerPool"....the worker pool logic is much simpler now, it's all in the queue handler and the queue set.

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1083,131 +997,140 @@ public boolean remove(Runnable task) {
+        TaskWrapper tw = it.next();
+        tw.cancel();
+        if (! (tw.task instanceof ShutdownRunnable)) {
+          int index = ListUtils.getInsertionEndIndex(resultList, tw, true);
```
I now added the feature that on shutdown the returned task list is ordered roughly in the order of execution....before it was basically a completely random list.

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1083,131 +997,140 @@ public boolean remove(Runnable task) {
-      synchronized (scheduleQueue.getModificationLock()) {
-        if (! workerPool.isShutdownStarted()) {
+    
+    public TaskWrapper getNextTask() {
```
Give a look at this logic...this is where we pick which task from each "QueueSet"...basically we figure out if we should have an execute task, or a scheduled task, which ever one is _most_ ready.

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
-                break;
-              default:
-                throw new UnsupportedOperationException();
+          worker = workerPool.getWorker();
```
Here is where we get the worker, then the next task, then provide the task to the worker...and loop.

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+     * @return Task to be executed next, or {@code null} if shutdown while waiting for task
+     * @throws InterruptedException Thrown if calling thread is interrupted while waiting for a task
+     */
+    protected TaskWrapper getNextReadyTask() throws InterruptedException {
```
Take a look at this logic...this is where we decide if we should run the low priority task or the high priority task...the first couple cases are if only one task is in queue...those will be rare.  The bottom case is where we have a task in both queues, and we find out which one is ready to execute, and if both are ready which one should be ran.

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1355,7 +1353,7 @@ protected void shutdownService() {
      * @param task Task to run on this workers thread
      */
-    public void nextTask(Runnable task) {
+    public void nextTask(TaskWrapper task) {
```
We have to take in a TaskWrapper here for the StatisticTracker

jentfoo commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1446,10 +1444,10 @@ public void cancel() {
+     * 
+     * @return true if the task should be executed
      */
-    public void executing() {
```
This is what checks to ensure that the task CAN execute.  If this returns false we loop for a new task.  This is why all tasks take in the queue, to handle that they are able to atomically be removed (and thus ran)....as well as handling the rescheduling into the queue when they are done.

jentfoo commented on src/main/java/org/threadly/concurrent/PrioritySchedulerStatisticTracker.java:
```diff
@@ -518,72 +549,6 @@ public int getCurrentRunningCount() {
   }
   
   /**
-   * Call to see how frequently tasks are able to immediately get a thread to execute on (and NOT 
```
I will have another pull request where these are marked as deprecated for a 3.7.0 release.

jentfoo commented on src/main/java/org/threadly/concurrent/PrioritySchedulerStatisticTracker.java:
```diff
@@ -676,114 +637,66 @@ protected static void trimWindow(Deque window) {
-            }
-          }
+    public void nextTask(TaskWrapper task) {
+      // may not be a wrapper for internal tasks like shutdown
```
Here is where we track how far delayed the task was...changes above needed to happen to ensure we could reliably check this here.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -765,64 +747,30 @@ public void prestartAllThreads() {
-     * This function REQUIRES that workersLock is synchronized before calling.  It returns an 
-     * available worker if it can get one before the wait time expires.  It will never create 
-     * a new worker.
+     * Call to get (or create, if able to) an idle worker.  If there are no idle workers available 
```
would this still create?

jentfoo replied:
Yes, this actually would not create before, but it does not.  That's why it is called "getWorker".  If a worker is not immediately available, and the pool is not maxed out, it will create one.  Because we get the worker before any tasks that means there will be one idle worker thread on creation, before any tasks are submitted to the scheduler (in part because we start the task consumer before any tasks are submitted to the pool as well).

We could delay these thread starts till task execution, but I did not see much reason, this pool is not really designed to be created and tossed away frequently.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1083,131 +997,140 @@ public boolean remove(Runnable task) {
+          if (scheduleDelay < executeDelay) {
+            return nextScheduledTask;
+          } else {
+            return nextExecuteTask;
```
so we favor execute tasks?

jentfoo replied:
Yes, we have to pick one to favor.  In general it should not make any sense because executed or scheduled tasks should not require accurate time execution.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1083,131 +997,140 @@ public boolean remove(Runnable task) {
-        if (! workerPool.isShutdownStarted()) {
+    
+    public TaskWrapper getNextTask() {
+      TaskWrapper nextScheduledTask = scheduleQueue.peekFirst();
```
i take it this is null if the next scheduled tasks delay is not up yet?

jentfoo replied:
Actually it would only be null if there are NO scheduled or executed tasks.  This function will return tasks which are not ready yet so that the queue manager knows how long it needs to park to wait for the task to become ready.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1083,131 +997,140 @@ public boolean remove(Runnable task) {
      */
-    public boolean isQueueEmpty() {
-      return executeQueue.isEmpty() && scheduleQueue.isEmpty();
+    public QueueSet getQueueSet(TaskPriority priority) {
```
this might be quicker as an if/else if/else.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
-              default:
-                throw new UnsupportedOperationException();
+          worker = workerPool.getWorker();
+          if (worker != null) {
```
have you messed around with switching the order here at all?  not sure why it would matter but might be worth a few benchmarks (grabbing the nextTask before the worker.

jentfoo replied:
I guess the reason I was grabbing the worker first now is because new tasks can be submitted (for example high priority) that would change the choosen task for execution while we wait on a worker.  This keeps the priority of execution accurate, while only having one thread to do the submissions of tasks to the worker threads.

That's the reason with multiple threads we could wait to grab the worker until we have a task ready to run.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+            // only high priority task is queued, so run if ready
+            long nextTaskDelay = nextHighPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+            if (nextTaskDelay <= 0) {
+              if (nextHighPriorityTask.canExecute()) {
```
would it make any sense to turn these 2 if into a while( && loop, might add some performance in the case where we run through a few tasks that exec out of under us. ( i assume that can happen here)

jentfoo replied:
It actually is in a while loop.  If you look at line 1187 (new line number), you will see we loop till we find a task to return, or until we shutdown the scheduler.

You are right, tasks can be removed (not because of executing, but because of a manual removal).  This is where we make sure we can automically remove the task for execution.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+          // only low priority task is queued, so run if ready
+          long nextTaskDelay = nextLowPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+          if (nextTaskDelay <= 0) {
+            if (nextLowPriorityTask.canExecute()) {
```
same as above with the while loop

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+          // both tasks are available, so see which one makes sense to run
+          long nextHighDelay;
+          long nextLowDelay;
+          ClockWrapper.stopForcingUpdate();
```
I get why your using the wrapper, I just wonder if it really provides much benefit here (there was another situation that was the same above).  I would say just order these in the way you want to weigh the bias of which is chosen to reduce this buy a few calls.

jentfoo replied:
We use the wrapper not for performance, but to prevent unfairly choosing (or inserting) one task incorrectly due to clock changes during inspection.  This is actually a deficency in the java.util.concurrent scheduler that we have solved.

I will play with the ordering of these a bit though.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+          }
+          if (nextHighDelay <= 0) {
+            // high task is ready, make sure it should be picked
+            if (nextHighDelay <= nextLowDelay || nextLowDelay > maxWaitForLowPriorityInMs * -1) {
```
the 2nd condition seems odd to me, you are basically forcing a delay on the low tasks as long as we have high tasks.  I assume thats the expected behavior, just seems like we should be doing that a little differently, like if there is any lowTasks with < 2x the delay of the current highTask then it needs to be ran.

jentfoo replied:
Here was the intention of this logic:

* If there is ever a high task that is delayed longer than a low task, execute the high task
* If there is a high task and a low task ready to run, and the above condition is not true, run the high task unless the low task has been waiting it's maximum wait time
* If there is just a high or just a low task ready to run, run it immediately

So the second condition is doing the second bullet, at that point we have both a high and a low priority task. But if the low priority task has not been waiting it's maximum wait time, we want to favor the high priority task.

Do you think it makes sense now?

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+            return nextLowPriorityTask;
+          } else {
+            if (nextLowDelay < nextHighDelay) {
+              LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextLowDelay);
```
seems like we could hint at what we expect to be ran in the next loop here.  possibly bypassing a lot of logic for the first loop.  Not sure exactly how it would work but could be worth looking into.

jentfoo replied:
Not sure exactly how we would do this.  We have no way to know why we woke up from a park, could be:
* We were unparked due to an update of one of the queues
* We were unparked due to reaching our timeout
* We were unparked due to the thread being interrupted
* We were unparked randomly

At one point I had the "QueueSet" remember it's last returned task as a cached value.  I was not able to get any extra performance from it, but maybe it's worth trying again.

lwahlmeier commented on src/main/java/org/threadly/concurrent/PriorityScheduler.java:
```diff
@@ -1539,18 +1549,25 @@ protected long getNextDelayInMillis() {
-        return;
+        return false;
+      }
+      synchronized (queueSet.scheduleQueue.getModificationLock()) {
```
I feel like this should be doable w/o the lock. assuming the remove is atomic and would only return true for the call that actually removes.

jentfoo replied:
The remove is atomic.  The reason we have to lock here is because if we have a recurring task, we need to make sure it is both removed, and re-added to the queue atomically.  Otherwise we could get into a condition where we may not be able to remove the recurring task because it can not be found in the queue.

That said, I will see if a reposition call is more efficent.

lwahlmeier commented:
All and all this looks pretty good.  I will go through the code once more tomorrow, so many changes its hard to just look at the diff and get a good sense whats happening.

jentfoo commented:
Alright @lwahlmeier....I did one final force push on this.  I updated some javadocs I had missed that were out of date with this behavior change.  I also moved the inner classes up above the WorkerPool.  Hopefully it does not make the diff more messy for you, but I think it's cleaner.

Can you give this one final look, and if it looks good to you, go ahead and merge it?

lwahlmeier commented:
So I have checked out the code.  I like it alot more then before, it seems to flow very well.

I still think we can reduce some code and simplify a few things, hopfully adding some performance, but I think for 4.0 this is pretty good.

jentfoo commented:
Awesome, thanks for taking a look.  I also have come to like it lot more too.  I think if we can do that TODO note on the QueueSet in PriorityScheduler (hopefully without creating a bunch of top level interfaces or classes) it will reduce a ton of code between the PriorityScheduler and the NoThreadScheduler.

I have been meaning to look into that again (and want to again before I release 4.X).

Since I am not sure exactly right now when we will release 4.X, we may get things even cleaner and faster before the release anyways (but it's good to get this massive pull request behind us).
